File: packages/diseno_flutter/lib/moleculas/chips_de_filtro.dart
import 'package:flutter/material.dart';

import '../atomos/chip_elegible.dart';
import '../tokens/tokens.dart';

/// Chips de filtro **en varias líneas**, cada uno con su ícono, sin «otros» (gate de F4).
class ChipsDeFiltro<T> extends StatelessWidget {
  const ChipsDeFiltro({
    super.key,
    required this.opciones,
    required this.valor,
    required this.onChanged,
  });
  final List<({T clave, String texto, IconData icono})> opciones;
  final T valor;
  final ValueChanged<T> onChanged;

  @override
  Widget build(BuildContext context) {
    return Wrap(
      spacing: Espacio.s2,
      runSpacing: Espacio.s2,
      children: [
        for (final o in opciones)
          ChipElegible(
            texto: o.texto,
            icono: o.icono,
            seleccionado: o.clave == valor,
            onTap: () => onChanged(o.clave),
          ),
      ],
    );
  }
}

File: packages/ui/src/chips-de-filtro/chips-de-filtro.ts
import { ChangeDetectionStrategy, Component, input, model } from '@angular/core'
import { ChipElegible } from '../chip-elegible/chip-elegible'
import type { NombreDeIcono } from '../icono/icono'

export type Filtro = { valor: string; texto: string; icono?: NombreDeIcono }

/**
 * Filtros de varias líneas, con ícono por chip y **sin «otros»**: cada tipo tiene su chip
 * o no se filtra. Selección múltiple; vacío = todos.
 */
@Component({
  selector: 'ap-chips-de-filtro',
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [ChipElegible],
  host: { role: 'group', '[attr.aria-label]': 'etiqueta()' },
  template: `
    @for (f of filtros(); track f.valor) {
      <ap-chip-elegible [icono]="f.icono" [elegido]="elegidos().includes(f.valor)" (elegidoChange)="alternar(f.valor, $event)">{{ f.texto }}</ap-chip-elegible>
    }
  `,
  styles: `:host { display: flex; flex-wrap: wrap; gap: var(--s2); }`,
})
export class ChipsDeFiltro {
  readonly etiqueta = input('Filtrar')
  readonly filtros = input.required<Filtro[]>()
  readonly elegidos = model<string[]>([])
  alternar(valor: string, elegido: boolean): void {
    const sin = this.elegidos().filter((v) => v !== valor)
    this.elegidos.set(elegido ? [...sin, valor] : sin)
  }
}
